### PR TITLE
Allow @Modules to be interfaces

### PR DESCRIPTION
**Pull request by JakeWharton**
Closes #13

### CODE REVIEW

File: reflect/src/main/java/dagger/reflect/ReflectiveModuleParser.java [hunk 1]
```diff
@@ -72,7 +84,12 @@ static void parse(Class<?> moduleClass, @Nullable Object instance,
 
         graphBuilder.add(key, binding);
       }
-      target = target.getSuperclass();
+
+      Class<?> superclass = target.getSuperclass();
+      if (superclass != Object.class && superclass != null) {
```
Comment (TWiStErRob) [inline reflect/src/main/java/dagger/reflect/ReflectiveModuleParser.java]: add `Object.class` and `null` to initial value of `seen` and let it `continue;`?

File: reflect/src/main/java/dagger/reflect/ReflectiveModuleParser.java [hunk 2]
```diff
@@ -72,7 +84,12 @@ static void parse(Class<?> moduleClass, @Nullable Object instance,
...
+        queue.add(superclass);
+      }
+      Collections.addAll(queue, target.getInterfaces());
```
Comment (TWiStErRob) [inline reflect/src/main/java/dagger/reflect/ReflectiveModuleParser.java]: would it worth testing multiple inheritance?
i.e. `abstract class Foo extends Bar implements Baz, Quz`

File: reflect/src/main/java/dagger/reflect/ReflectiveModuleParser.java (hunk 1 above)
Comment (JakeWharton) [reply reflect/src/main/java/dagger/reflect/ReflectiveModuleParser.java]: I'd prefer to keep these out of the collections altogether since they're more like bad data

File: reflect/src/main/java/dagger/reflect/ReflectiveModuleParser.java (hunk 2 above)
Comment (JakeWharton) [reply reflect/src/main/java/dagger/reflect/ReflectiveModuleParser.java]: Added a test for this by way of testing that when an interface is included twice via a base class that it still works

File: reflect/src/main/java/dagger/reflect/ReflectiveModuleParser.java (hunk 1 above)
Comment (TWiStErRob) [reply reflect/src/main/java/dagger/reflect/ReflectiveModuleParser.java]: edge case: if the method is called with either of these, it can get inside the collections, and the error won't be graceful

File: integration-tests/src/main/java/com/example/ModuleClassHierarchy.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+  @Module
+  abstract class Module1 extends BaseModule {
+    @Provides static String string() {
+      return "foo";
+    }
+  }
+
+  @Module
+  abstract class BaseModule {
```
Comment (TWiStErRob) [inline integration-tests/src/main/java/com/example/ModuleClassHierarchy.java]: I think the `@Module` annotation on the base classes is not necessary, because it's not used in `modules =`

File: reflect/src/main/java/dagger/reflect/ReflectiveModuleParser.java (hunk 2 above)
Comment (TWiStErRob) [reply reflect/src/main/java/dagger/reflect/ReflectiveModuleParser.java]: I was going for more the multiple interface to make sure all of the interfaces are checked on a single class not just the first one, but up to you

File: integration-tests/src/main/java/com/example/ModuleClassHierarchy.java (hunk 3 above)
Comment (JakeWharton) [reply integration-tests/src/main/java/com/example/ModuleClassHierarchy.java]: The compiler requires all types with providers to contain it

File: integration-tests/src/test/java/com/example/IntegrationTest.java [hunk 4]
```diff
@@ -256,6 +256,43 @@
...
+  }
+
+  @Test public void moduleInterfaceHierarchy() {
+    ModuleInterface component = backend.create(ModuleInterface.class);
+    assertThat(component.string()).isEqualTo("foo");
+  }
+
   private void ignoreReflectionBackend() {
     assumeTrue("Not yet implemented for reflection backend", backend != Backend.REFLECT);
   }
```
Comment (TWiStErRob) [inline integration-tests/src/test/java/com/example/IntegrationTest.java]: the length of this file and package is getting scary, have you thought about splitting it to multiple tests and moving the input classes to packages? (e.g. this batch of new tests would go into `com.example.module_hierarchy` or similar)

File: reflect/src/main/java/dagger/reflect/ReflectiveModuleParser.java (hunk 1 above)
Comment (JakeWharton) [reply reflect/src/main/java/dagger/reflect/ReflectiveModuleParser.java]: Sure but that's parameter validation which is a separate concern. If you pass null you get an NPE and if you pass `Object.class` you get an error that its methods don't have `@Provides`

File: integration-tests/src/test/java/com/example/IntegrationTest.java (hunk 4 above)
Comment (JakeWharton) [reply integration-tests/src/test/java/com/example/IntegrationTest.java]: Not a priority to me, no. The methods are small and navigable enough with cmd+f12 for now
